joomla/joomla-cms
# [4.0] Controller classes improvements

Pull Request for Issue # .

### Summary of Changes
This PR makes several improvements to controller classes as explained below:

1. Add $option property to Controller class. We will now get $option from $config or input, not guessing it from class name like in none-namespace class anymore

2. Correct behavior or getName() method:
- For base component controller, it will return name controller (for example, contact, banners, modules)
- For sub-controller class, it will return the name of that sub-controller. For example, BannersControllerBanner will return banner.

3. Implement getModel method for admin controller. With this method, we can delete most of blank getModel method in admin controllers (I deleted some in this PR for demo purpose)

4. Some code clean up, docblock, codestyle for easy navigating between method/class in modern IDE like PHPStorm

5. Pass $option and $name property to Model and View creation code so that the Model and View class won't have to guess these important properties from class name anymore

### Testing Instructions
1. Install Joomla 4 from 4.0-dev branch or nightly build
2. Apply patch
3. Try to navigate to random pages in frontend and backend, make sure it still works as expected
4. Try to publish/unpublish banners for example, make sure it is working as expected

### Documentation Changes Required
1. getName() method of Controller return name of actual controller (not always return name of component like before)

2. Note the getModel method of Admin controller implemented.
